Reqnroll/CucumberMessages/Configuration/CucumberConfiguration.cs

```diff
@@ -0,0 +1,148 @@
+{
+    /// <summary>
+    /// This class is responsible for determining the configuration of the Cucumber Messages subsystem.
+    /// It is wired into the object container as a singleton and is a dependency of the PubSub classes.
+    /// 
+    /// When any consumer of this class asks for one of the properties of ICucumberConfiguration,
+    /// the class will resolve the configuration (only once).
+    /// 
+    /// A default configuration is provided (by DefaultConfigurationSource). 
+    /// It is supplemented by one or more profiles from the configuration file. (ConfigFile_ConfigurationSource)
+    /// Then Environmment Variable Overrides are applied.
+    /// </summary>
+    public class CucumberConfiguration : ICucumberMessagesConfiguration
+    {
+        public static ICucumberMessagesConfiguration Current { get; private set; }
+        public bool Enabled => _enablementOverrideFlag && _resolvedConfiguration.Value.Enabled;
+        public string BaseDirectory => _resolvedConfiguration.Value.BaseDirectory;
+        public string OutputDirectory => _resolvedConfiguration.Value.OutputDirectory;
+        public string OutputFileName => _resolvedConfiguration.Value.OutputFileName;
+        public IDGenerationStyle IDGenerationStyle => _resolvedConfiguration.Value.IDGenerationStyle;
+
+        private readonly IObjectContainer _objectContainer;
+        private Lazy<ITraceListener> _traceListenerLazy;
+        private IEnvironmentWrapper _environmentWrapper;
+        private IReqnrollJsonLocator _reqnrollJsonLocator;
+        private Lazy<ResolvedConfiguration> _resolvedConfiguration;
+        private bool _enablementOverrideFlag = true;
+
+        public CucumberConfiguration(IObjectContainer objectContainer, IEnvironmentWrapper environmentWrapper, IReqnrollJsonLocator configurationFileLocator)
+        {
+            _objectContainer = objectContainer;
+            _traceListenerLazy = new Lazy<ITraceListener>(() => _objectContainer.Resolve<ITraceListener>());
+            _environmentWrapper = environmentWrapper;
+            _reqnrollJsonLocator = configurationFileLocator;
+            _resolvedConfiguration = new Lazy<ResolvedConfiguration>(ResolveConfiguration);
+            Current = this;
+        }
+
+        #region Override API
+        public void SetEnabled(bool value)
+        {
+            _enablementOverrideFlag = value;
+        }
+        #endregion
+
+
+        private ResolvedConfiguration ResolveConfiguration()
+        {
+            var config = ApplyHierarchicalConfiguration();
+            var resolved = ApplyEnvironmentOverrides(config);
+
+            // a final sanity check, the filename cannot be empty
+            if (string.IsNullOrEmpty(resolved.OutputFileName))
+            {
+                resolved.OutputFileName = "reqnroll_report.ndjson";
+                _traceListenerLazy.Value.WriteToolOutput($"WARNING: Cucumber Messages: Output filename was empty. Setting filename to {resolved.OutputFileName}");
```

Comment:
I don't think we should provide a warning here. If a config value is optional, it is valid to leave it as default, so we should not issue a warning.

Reply:
Removed.